sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -603,6 +603,19 @@ case class AdaptiveSparkPlanExec(
     (newPlan, optimized)
   }
 
+  /**
+   * Clean up logical plan stats before re-optimize
+   */
+  private def cleanupStats(logicalPlan: LogicalPlan): Unit = {
+    logicalPlan.invalidateStatsCache()
+    // We must invalidate ineffective rules before re-optimize since AQE Optimizer may introduce
+    // LocalRelation that can affect result.
```

Comment:
Why do we need to invalidate ineffective rules?  If AQE Optimizer rewrote a leaf node to `LocalRelation`,  the path from the new `LocalRelation` to the new root are all *new* TreeNode instances, in which ineffective rule bits are not set.

Reply:
Let's say we have a complex plan.
* two join with two exchange
*  join2 rigth side is empty

the logical join looks like:
```
Join2
:- Join1
:  :  +- Relation parquet
:  +- LogicalQueryStage Project, BroadcastQueryStage 0
+- LogicalQueryStage Relation parquet, ShuffleQueryStage 1
```

* first round, QueryStage 0 is materialized and QueryStage 1 is not materialized. Then AQE Optimier can only optimize join1 but unfortunately the rule has no effects. At this monent, we would mark this rule as ineffictive.
* second round, QueryStage 1 is materialized. Then AQE Optimier can optimize join2 but the rule has been already marked as ineffictive. As a result we cann't optimize it.

Reply:
Ok, IIUC, wether a stage is "materialized or not" is kept as an external varying state outside of plan nodes?  
If that's the case, the same rule object is invoked multiple times for the same logical plan but violates the contract of passing `ruleId`s:
https://github.com/databricks/runtime/blob/88acfbbb14f1396c312c394ed8a2a645738f83f0/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala#L445-L446

I'd prefer us keeping ineffective rule bits internal to TreeNodes for simplicity, otherwise it might be difficult to reason about certain behaviors.  I suspect we don't have that many fixpoint iterations in AQEOptimizer? What `ruleId` helped most are Analyzer rules, because the fix-point batch can run 5~8 iterations.

Reply:
Seems it's meaningless to do `transformUpWithPruning` at AQE side. I see what you worried but there has an another issue:

Now we reuse the rule which is already used `transformUpWithPruning` between normal and AQE Optimizer and maybe we will introduce other rules in future. Then If we don't invaildate ineffective rules, we need to make new rule for every exists rule. or we can use other tag to distinguish where the rule used ?

Reply:
>> Now we reuse the rule which is already used transformUpWithPruning 
>> between normal and AQE Optimizer and maybe we will introduce other rules in future.

Can you elaborate a bit more?  I'm not sure if I understand the exact issue here. 
IIUC, the main issue here is that `invalidateStatsCache` breaks the `immutability` property of LogicalPlans and SparkPlans?

Reply:
It's not about `invalidateStatsCache`. The materialized flag of logical plan will be changed during running and some AQE Optimier rules only match the materialized logical plan so we cann't promise whether a rule is ineffective or not before plan materialized.

Reply:
I think it makes sense to clear the rule id bits and rerun all the optimization rules, which is the same as what we did before the tree node pattern framework.

Reply:
Make sense, don't use ruleId to do transform if the plan is not immutable. thus, let me leave some comments about this.